ipn/ipnlocal, control/controlclient: data race

```
=== RUN   TestFunnelClose
=== RUN   TestFunnelClose/simple
    tsnet_test.go:211: DERP httpsrv listener: 127.0.0.1:37831
    tsnet_test.go:224: testcontrol listening on http://127.0.0.1:44773/
2026/06/01 19:21:06 tsnet starting with hostname "s", varRoot "/tmp/TestFunnelClosesimple3346185409/001/s"
2026/06/01 19:21:06 LocalBackend state is NeedsLogin; running StartLoginInteractive...
    testcontrol.go:368: Created user userid:1 for node nodekey:db8462662b971343b028d3eb481c79a1824537cf64b093943bddbcf24803826c
2026/06/01 19:21:06 AuthLoop: state is Starting; done
    stuntest.go:65: STUN server shutdown
--- PASS: TestFunnelClose/simple (0.58s)
=== RUN   TestFunnelClose/no_clobbering
    tsnet_test.go:211: DERP httpsrv listener: 127.0.0.1:44623
    tsnet_test.go:224: testcontrol listening on http://127.0.0.1:46329/
2026/06/01 19:21:07 tsnet starting with hostname "s", varRoot "/tmp/TestFunnelCloseno_clobbering2607587850/001/s"
2026/06/01 19:21:07 LocalBackend state is NeedsLogin; running StartLoginInteractive...
    testcontrol.go:368: Created user userid:1 for node nodekey:7ed255086cd712d36c1102998b91864f051c0dbe0ccc5cd1476f43cc01bbac5f
2026/06/01 19:21:07 AuthLoop: state is Starting; done
    stuntest.go:65: STUN server shutdown
--- PASS: TestFunnelClose/no_clobbering (0.59s)
=== RUN   TestFunnelClose/two_listeners
    tsnet_test.go:211: DERP httpsrv listener: 127.0.0.1:34549
    tsnet_test.go:224: testcontrol listening on http://127.0.0.1:40663/
2026/06/01 19:21:07 tsnet starting with hostname "s1", varRoot "/tmp/TestFunnelClosetwo_listeners2091133677/001/s1"
2026/06/01 19:21:07 LocalBackend state is NeedsLogin; running StartLoginInteractive...
    testcontrol.go:368: Created user userid:1 for node nodekey:be59cb0aeb0b8b75f9e16f0d5753db14330509e040cd6b9fa790bd0ae206f726
2026/06/01 19:21:07 AuthLoop: state is Starting; done
==================
WARNING: DATA RACE
Write at 0x00c0027b2ac0 by goroutine 1443:
  tailscale.com/control/controlclient.(*Auto).cancelMapCtxLocked()
      /home/runner/work/tailscale/tailscale/src/control/controlclient/auto.go:290 +0xbc
  tailscale.com/control/controlclient.(*Auto).SetPaused()
      /home/runner/work/tailscale/tailscale/src/control/controlclient/auto.go:221 +0x20b
  tailscale.com/ipn/ipnlocal.(*LocalBackend).pauseOrResumeControlClientLocked()
      /home/runner/work/tailscale/tailscale/src/ipn/ipnlocal/local.go:1072 +0xb0
  tailscale.com/ipn/ipnlocal.(*LocalBackend).enterStateLocked()
      /home/runner/work/tailscale/tailscale/src/ipn/ipnlocal/local.go:6452 +0x605
  tailscale.com/ipn/ipnlocal.(*LocalBackend).stateMachineLocked()
      /home/runner/work/tailscale/tailscale/src/ipn/ipnlocal/local.go:6606 +0x35
  tailscale.com/ipn/ipnlocal.(*LocalBackend).setPrefsLocked()
      /home/runner/work/tailscale/tailscale/src/ipn/ipnlocal/local.go:5330 +0x12f1
  tailscale.com/ipn/ipnlocal.(*LocalBackend).editPrefsLocked()
      /home/runner/work/tailscale/tailscale/src/ipn/ipnlocal/local.go:5220 +0x80d
  tailscale.com/ipn/ipnlocal.(*LocalBackend).Logout()
      /home/runner/work/tailscale/tailscale/src/ipn/ipnlocal/local.go:6750 +0x20d
  tailscale.com/ipn/ipnlocal.(*LocalBackend).Shutdown()
      /home/runner/work/tailscale/tailscale/src/ipn/ipnlocal/local.go:1284 +0x2f1
  tailscale.com/tsnet.(*Server).close()
      /home/runner/work/tailscale/tailscale/src/tsnet/tsnet.go:651 +0x424
  tailscale.com/tsnet.(*Server).Close.func1()
      /home/runner/work/tailscale/tailscale/src/tsnet/tsnet.go:606 +0x45
  sync.(*Once).doSlow()
      /home/runner/.cache/tsgo/7104161f96b07474ce312ea116c8b2a1efe1311a/src/sync/once.go:78 +0xa1
  tailscale.com/tsnet.startServer.func1()
      /home/runner/work/tailscale/tailscale/src/tsnet/tsnet_test.go:331 +0x2e
  testing.(*common).Cleanup.func1()
      /home/runner/.cache/tsgo/7104161f96b07474ce312ea116c8b2a1efe1311a/src/testing/testing.go:1317 +0x168
  testing.(*common).runCleanup()
      /home/runner/.cache/tsgo/7104161f96b07474ce312ea116c8b2a1efe1311a/src/testing/testing.go:1667 +0x225
  testing.tRunner.func2()
      /home/runner/.cache/tsgo/7104161f96b07474ce312ea116c8b2a1efe1311a/src/testing/testing.go:2030 +0x4c
  runtime.deferreturn()
      /home/runner/.cache/tsgo/7104161f96b07474ce312ea116c8b2a1efe1311a/src/runtime/panic.go:668 +0x5d
  testing.(*T).Run.gowrap1()
      /home/runner/.cache/tsgo/7104161f96b07474ce312ea116c8b2a1efe1311a/src/testing/testing.go:2101 +0x38

Previous read at 0x00c0027b2ac0 by goroutine 1542:
  tailscale.com/control/controlclient.mapRoutineState.UpdateUserProfiles()
      /home/runner/work/tailscale/tailscale/src/control/controlclient/auto.go:543 +0x16e
  tailscale.com/control/controlclient.(*mapRoutineState).UpdateUserProfiles()
      <autogenerated>:1 +0x4e
  tailscale.com/control/controlclient.(*mapSession).tryHandleIncrementally()
      /home/runner/work/tailscale/tailscale/src/control/controlclient/map.go:445 +0x549
  tailscale.com/control/controlclient.(*mapSession).handleNonKeepAliveMapResponse()
      /home/runner/work/tailscale/tailscale/src/control/controlclient/map.go:334 +0x868
  tailscale.com/control/controlclient.(*mapSession).run()
      /home/runner/work/tailscale/tailscale/src/control/controlclient/map.go:154 +0x104
  tailscale.com/control/controlclient.newMapSession.gowrap1()
      /home/runner/work/tailscale/tailscale/src/control/controlclient/map.go:140 +0x2e

Goroutine 1443 (running) created at:
  testing.(*T).Run()
      /home/runner/.cache/tsgo/7104161f96b07474ce312ea116c8b2a1efe1311a/src/testing/testing.go:2101 +0xb12
  tailscale.com/tsnet.TestFunnelClose()
      /home/runner/work/tailscale/tailscale/src/tsnet/tsnet_test.go:892 +0x17a
  testing.tRunner()
      /home/runner/.cache/tsgo/7104161f96b07474ce312ea116c8b2a1efe1311a/src/testing/testing.go:2036 +0x21c
  testing.(*T).Run.gowrap1()
      /home/runner/.cache/tsgo/7104161f96b07474ce312ea116c8b2a1efe1311a/src/testing/testing.go:2101 +0x38

Goroutine 1542 (running) created at:
  tailscale.com/control/controlclient.newMapSession()
      /home/runner/work/tailscale/tailscale/src/control/controlclient/map.go:140 +0x58a
  tailscale.com/control/controlclient.(*Direct).sendMapRequest()
      /home/runner/work/tailscale/tailscale/src/control/controlclient/direct.go:1252 +0x2a77
  tailscale.com/control/controlclient.(*Direct).PollNetMap()
      /home/runner/work/tailscale/tailscale/src/control/controlclient/direct.go:966 +0x453
  tailscale.com/control/controlclient.(*Auto).mapRoutine()
      /home/runner/work/tailscale/tailscale/src/control/controlclient/auto.go:621 +0x3b7
  tailscale.com/control/controlclient.(*Auto).start.gowrap3()
      /home/runner/work/tailscale/tailscale/src/control/controlclient/auto.go:247 +0x2e
==================
    stuntest.go:65: STUN server shutdown
    testing.go:1712: race detected during execution of test
--- FAIL: TestFunnelClose/two_listeners (0.61s)
```

/cc @tailscale/client-core

--- a/control/controlclient/auto.go
+++ b/control/controlclient/auto.go
@@ -472,13 +472,14 @@ func (mrs mapRoutineState) UpdateNetmapDelta(muts []netmap.NodeMutation) bool {
 	c.mu.Lock()
 	goodState := c.loggedIn && c.inMapPoll
 	ndu, canDelta := c.observer.(NetmapDeltaUpdater)
+	mapCtx := c.mapCtx
 	c.mu.Unlock()
 
 	if !goodState || !canDelta {
 		return false
 	}
 
-	ctx, cancel := context.WithTimeout(c.mapCtx, 2*time.Second)
+	ctx, cancel := context.WithTimeout(mapCtx, 2*time.Second)
 	defer cancel()
 
 	ch := make(chan bool, 1)
@@ -508,11 +509,12 @@ func (mrs mapRoutineState) UpdatePacketFilter(rules views.Slice[tailcfg.FilterRu
 	c.mu.Lock()
 	goodState := c.loggedIn && c.inMapPoll
 	pfu, ok := c.observer.(PacketFilterUpdater)
+	mapCtx := c.mapCtx
 	c.mu.Unlock()
 	if !goodState || !ok {
 		return false
 	}
-	ctx, cancel := context.WithTimeout(c.mapCtx, 2*time.Second)
+	ctx, cancel := context.WithTimeout(mapCtx, 2*time.Second)
 	defer cancel()
 	ch := make(chan bool, 1)
 	c.observerQueue.Add(func() {
@@ -536,11 +538,12 @@ func (mrs mapRoutineState) UpdateUserProfiles(profiles map[tailcfg.UserID]tailcf
 	c.mu.Lock()
 	goodState := c.loggedIn && c.inMapPoll
 	upu, ok := c.observer.(UserProfileUpdater)
+	mapCtx := c.mapCtx
 	c.mu.Unlock()
 	if !goodState || !ok {
 		return false
 	}
-	ctx, cancel := context.WithTimeout(c.mapCtx, 2*time.Second)
+	ctx, cancel := context.WithTimeout(mapCtx, 2*time.Second)
 	defer cancel()
 	ch := make(chan bool, 1)
 	c.observerQueue.Add(func() {
@@ -561,13 +564,14 @@ func (mrs mapRoutineState) PatchDiscoKey(pub key.NodePublic, disco key.DiscoPubl
 	c.mu.Lock()
 	goodState := c.loggedIn && c.inMapPoll
 	dun, ok := c.observer.(patchDiscoKeyer)
+	mapCtx := c.mapCtx
 	c.mu.Unlock()
 
 	if !goodState || !ok {
 		return
 	}
 
-	ctx, cancel := context.WithTimeout(c.mapCtx, 2*time.Second)
+	ctx, cancel := context.WithTimeout(mapCtx, 2*time.Second)
 	defer cancel()
 
 	c.observerQueue.RunSync(ctx, func() {

--- a/control/controlclient/auto_test.go
+++ b/control/controlclient/auto_test.go
@@ -1,0 +1,66 @@
+// Copyright (c) Tailscale Inc & contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+package controlclient
+
+import (
+	"context"
+	"sync"
+	"testing"
+	"time"
+
+	"tailscale.com/tailcfg"
+)
+
+type userProfileUpdateObserver struct{}
+
+func (userProfileUpdateObserver) SetControlClientStatus(Client, Status) {}
+
+func (userProfileUpdateObserver) UpdateUserProfiles(map[tailcfg.UserID]tailcfg.UserProfileView) bool {
+	return true
+}
+
+func TestMapRoutineStateUpdateUserProfilesConcurrentCancelMapCtx(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	c := &Auto{
+		logf:      func(string, ...any) {},
+		observer:  userProfileUpdateObserver{},
+		mapCtx:    ctx,
+		mapCancel: cancel,
+		loggedIn:  true,
+		inMapPoll: true,
+	}
+	mrs := mapRoutineState{c: c}
+
+	start := make(chan struct{})
+	var wg sync.WaitGroup
+	for range 4 {
+		wg.Go(func() {
+			<-start
+			for range 2000 {
+				c.mu.Lock()
+				c.cancelMapCtxLocked()
+				c.mu.Unlock()
+			}
+		})
+	}
+	for range 4 {
+		wg.Go(func() {
+			<-start
+			for range 2000 {
+				mrs.UpdateUserProfiles(nil)
+			}
+		})
+	}
+
+	close(start)
+	wg.Wait()
+
+	waitCtx, cancel := context.WithTimeout(t.Context(), 5*time.Second)
+	defer cancel()
+	if err := c.observerQueue.Wait(waitCtx); err != nil {
+		t.Fatal(err)
+	}
+	c.observerQueue.Shutdown()
+	c.mapCancel()
+}

